nvaccess/nvda
# Fix currently edited profile announcements, numpad enter functionality and validation behavior

### Link to issue number:
Fixes #8264 
Fixes #8310 

### Edited profile enhancements
#### Steps to reproduce
1. Focus an application with a custom configuration profile activated
2. Open NVDA's settings dialog using the menu
3. Press down arrow to go to speech settings

#### Expected result
NVDA says: "Editing profile <profile_name>"

#### Actual result
NVDA does not announce the profile name on the first switch. Subsequent switches are ok.

I've actually seen this earlier and thought I had fixed it properly, but it turns out that I only tested this by means of opening the settings dialog with NVDA+control+g, in which case this issue can not be reproduced. This is because event_appModule_gainFocus is only triggered when the NVDA process gets the foreground, not when going from the menu to a settings dialog.

#### Description of how this pull request fixes the issue:
This moves the logic that was in the NVDA appModule_gainFocus event handler to the event_foreground handler, which is actually much more elegant. It also seemed possible to remove event_appModule_loseFocus as well, however a more detailed look at the speech viewer revealed that NVDA silenced itself (i.e. there were unwanted profile editing change notifications that I didn't hear because another speech message overruled them).

#### Testing performed:
Tested using the steps to reproduce,, both by opening the settings dialog by means of gestures and using the NVDA menu.

Also tested:

1. Open the dialog with NVDA+control+d, custom profile active.
2. Go to the general category. NVDA announces the switch
3. Close the dialog
4. Re-open the dialog with NVDA+control+d, custom profile active.
5. NVDA only announces the active profile as part of the title bar, which is expected. I.e. NVDA should not announce the "editing profile" message.

### Numpad enter functionality
The _enterTriggersOnOk_ctrlSTriggersOnApply handler now also triggers OK when pressing numpad enter.

### onSave popups
In the old situation, onSave for the general and keyboard panels could pop-up a warning or error, respectively. This has now been fixed by creating two additional handlers on panels, preSave and postSave. preSave is evaluated for every panel before settings are actually saved. A preSave handler may either return True or false. If False, the preSave handler blocks saving of all settings in the multi category settings dialog.
the return value of postSave handlers is ignored for now, since there is currently no use case scenario for them blocking closing the multi category settings dialog, for example.

### Known issues with pull request:
None

### Change log entry:
None
